### PR TITLE
4140 - Implement return of funds when an offer is declined (modal + flash message) & 4544 - BUG: HoloFuel: debits are showing up as credits in "Activity" view of Inbox & 9284 - Diagnose Bugs and page wait-times in Holofuel & 6156 - I receive visual feedback on the action I took in Inbox.

### PR DESCRIPTION
**Pull request by JettTech**
- 4140: Adds new gql Mutation for bulk cancel of declined offers
- 4140: Updates Inbox with new modal and Flash Message for handling cancellation of declined offers
- 4140: Updates Primary Layout to catch and redirect to inbox, upon discovery of an incoming declined offer
- 4140: Updates `test:preseed-data` script with better setup/prep for live data testing of decline & cancel
- 4544: Corrects visual display for credits/debits in Inbox Activity view (resolves bug)
- 9284: Replaces page specific whois queries with a whois function, called for each tx returned from a fetch all transactions query type
- 6156 - Adds designs for visual feedback on actions completed in Inbox.

### CODE REVIEW

File: scripts/initialize-conductor-data.js [hunk 1]
```diff
@@ -115,9 +223,9 @@ async function populateHoloFuelData () {
 populateHoloFuelData()
```
Comment (zo-el) [inline scripts/initialize-conductor-data.js:222]: Can we have these scripts consistent and not change it depending on the test we need to run locally.

**Review by zo-el — CHANGES_REQUESTED** [review]
Can we delete the unused Queries? Like `HolofuelHomeCounterparties`...

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 2]
```diff
@@ -21,7 +20,14 @@ export async function getTxCounterparties (transactionList) {
   return noDuplicatesAgentList
 }
 
-const presentRequest = ({ origin, event, stateDirection, eventTimestamp, counterpartyId, amount, notes, fees, status, reason }) => {
+const addFullCounterpartyToTx = async (tx) => {
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js:23]: I love this change, I had been considering this as an alternative to the useWithCounterparties hooks we've been using.

I'm curious if there's anything specific about these features that made you do it in this PR?

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 3]
```diff
@@ -230,25 +234,28 @@ const HoloFuelDnaInterface = {
       const { transactions } = await createZomeCall('transactions/list_transactions')()
       const listOfNonActionableTransactions = transactions.map(presentTransaction)
       const noDuplicateIds = _.uniqBy(listOfNonActionableTransactions, 'id')
+      const displayReadyCompletedTransactions = await getTxWithCounterparties(noDuplicateIds.filter(tx => tx.status === 'completed'))
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: this is personal preference feel free to ignore, but I would call these
`presentedCompletedTransactions`

`displayReadyCompletedTransactions` is a little harder to read because when you start reading it it seems like a verb.

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 4]
```diff
@@ -230,25 +234,28 @@ const HoloFuelDnaInterface = {
...
 
-      return noDuplicateIds.filter(tx => tx.status === 'completed').sort((a, b) => a.timestamp > b.timestamp ? -1 : 1)
+      return displayReadyCompletedTransactions.sort((a, b) => a.timestamp > b.timestamp ? -1 : 1)
     },
     allActionable: async () => {
       const { requests, promises, declined, canceled } = await createZomeCall('transactions/list_pending')()
-      const actionableTransactions = await requests.map(r => presentPendingRequest(r)).concat(promises.map(p => presentPendingOffer(p))).concat(declined.map(presentDeclinedTransaction)).concat(canceled.map(presentCanceledTransaction))
+      const actionableTransactions = await requests.map(r => presentPendingRequest(r)).concat(promises.map(p => presentPendingOffer(p))).concat(declined.map(presentDeclinedTransaction)).concat(canceled.map(presentIncomingCanceledTransaction))
+      const displayReadyActionableTransactions = await getTxWithCounterparties(actionableTransactions)
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: same, `presentedActionableTransactions`

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 5]
```diff
@@ -230,25 +234,28 @@ const HoloFuelDnaInterface = {
       const { transactions } = await createZomeCall('transactions/list_transactions')()
       const listOfNonActionableTransactions = transactions.map(presentTransaction)
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: I know this is not part of the current PR, but just noticing this. for naming arrays, we can just use a plural noun. Don't need to specify `listOf` in the name. So this can just be `nonActionableTransactions`

Also, given we call the same transactions `completed` later in the function, it's easier to follow if this is `completedTransactions`

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 6]
```diff
@@ -274,7 +281,17 @@ const HoloFuelDnaInterface = {
...
       const noDuplicateIds = _.uniqBy(listOfNonActionableTransactions, 'id')
 
-      return noDuplicateIds.filter(tx => tx.status !== 'pending').sort((a, b) => a.timestamp > b.timestamp ? -1 : 1)
+      const whoami = await HoloFuelDnaInterface.user.get()
+      const nonActionableTransactionsWithCancelByKey = noDuplicateIds
+        .filter(tx => tx.status !== 'pending')
+        .map(tx => {
+          if (tx.status === STATUS.canceled) return { ...tx, canceledBy: whoami }
+          else return { ...tx, canceledBy: null }
+        })
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: not blocking, but this could be a one line ternary
```
.map(tx -> tx.status === STATUS.cancelled ? { ...tx, canceledBy: whoami } : { ...tx, canceledBy: null }
```

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 7]
```diff
@@ -303,35 +325,47 @@ const HoloFuelDnaInterface = {
...
+        ...transaction,
+        id: transactionId
+      }
+    },
+    /* NOTE: recover funds from DECLINED PENDING TRANSACTION (ie: Counterparty declined offer) - intended for REFUNDS  */
+    recoverFunds: async (transactionId) => {
+      // const TX_DECLINED = 'declined-transaction'
+      const { rawTransaction, transaction } = await HoloFuelDnaInterface.transactions.getPendingDeclined(transactionId, { raw: true })
+      const canceledProof = await createZomeCall('transactions/cancel')({ entry: rawTransaction })
+      if (!canceledProof) throw new Error('Recover Funds Error.', canceledProof)
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: Couple of things
1) Error doesn't work like console.log, only the first param is used in the message. So to achieve this effect you would do
```
throw new Error(`Recover Funds Error. ${ canceledProof}`)
```
Easy mistake to make, I do this all the time.

But also
2) no point in including `canceledProof` in the error as it is always going to be null. As a result I think we can probably give a more specific error message. 
```
`Couldn't find a transaction with id ${transactionId}`
```

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 8]
```diff
@@ -303,35 +325,47 @@ const HoloFuelDnaInterface = {
...
-    return {
-      ...transaction,
-      id: transactionId,
-      status: STATUS.canceled,
-      canceledReference: canceledProof
+    },
+    refundAllDeclined: async (listOfDeclinedTransactions) => {
+      const listOfTransactionIds = listOfDeclinedTransactions.map(({ id }) => id)
+      const canceledProof = await createZomeCall('transactions/cancel_transactions')({ origins: listOfTransactionIds })
+      if (!canceledProof) throw new Error('Cancel Error.', canceledProof)
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: see above comment

File: src/data-interfaces/HoloFuelDnaInterface.js [hunk 9]
```diff
@@ -303,35 +325,47 @@ const HoloFuelDnaInterface = {
...
-    const reason = annulTransactionReason
-    const transaction = await HoloFuelDnaInterface.transactions.getPendingDeclined(transactionId)
-    const canceledProof = await createZomeCall('transactions/cancel')({ entry: transaction, reason })
-    return {
-      ...transaction,
-      id: transactionId,
-      status: STATUS.canceled,
-      canceledReference: canceledProof
+    },
+    refundAllDeclined: async (listOfDeclinedTransactions) => {
```
Comment (robbiecarlton) [inline src/data-interfaces/HoloFuelDnaInterface.js]: this name is misleading. `refundAllDeclined` to me says it refunds ALL declined transactions in the dna. I would just call it `refundDeclined`

File: src/graphql-server/schema.graphql [hunk 10]
```diff
@@ -57,7 +57,7 @@ enum TransactionDirection {
 }
 
 type Transaction {
-  id: ID # ID will be the originCommitHash
+  id: ID # ID will be the origin commit hash
```
Comment (robbiecarlton) [inline src/graphql-server/schema.graphql:60]: 😁👍

File: src/graphql-server/schema.graphql [hunk 11]
```diff
@@ -106,6 +107,23 @@ type HposAuth {
...
+  amount: Float
+  counterparty: HolofuelUserInput
+  direction: TransactionDirection
+  status: TransactionStatus
+  type: TransactionType
+  timestamp: String
+  fees: Float
+  notes: String
+}
+
```
Comment (robbiecarlton) [inline src/graphql-server/schema.graphql]: I don't think we need this Type.

See comment on `holofuelRefundAllDeclined` below

File: src/graphql-server/schema.graphql [hunk 12]
```diff
@@ -139,7 +157,8 @@ type Mutation {
   holofuelAcceptOffer(transactionId: ID): Transaction
   holofuelDecline(transactionId: ID): Transaction
   holofuelCancel(transactionId: ID): Transaction
-  holofuelRecoverFunds(transaction: ID): Transaction
+  holofuelRecoverFunds(transactionId: ID): Transaction
+  holofuelRefundAllDeclined(listOfDeclinedTransactions: [CancelDeclinedTransactionsInput]): [Transaction]
```
Comment (robbiecarlton) [inline src/graphql-server/schema.graphql]: I think we should just be passing a list of `IDs` here. We're not using any of the other data in `CancelDeclinedTransactionsInput`.

Also, `listOf` variable naming. How about:

`holofuelRefundAllDeclined(transactionIds: [ID]): [Transaction]`

File: src/graphql/HolofuelAcceptOfferMutation.gql [hunk 13]
```diff
@@ -2,7 +2,7 @@ mutation HolofuelAcceptOfferMutation ($transactionId: ID) {
   holofuelAcceptOffer(transactionId: $transactionId) {
     id
     amount
-    counterparty
+    counterparty { id, nickname }
```
Comment (robbiecarlton) [inline src/graphql/HolofuelAcceptOfferMutation.gql:5]: what do we use this value for?

File: src/graphql/HolofuelRefundAllDeclinedMutation.gql [hunk 14]
```diff
@@ -0,0 +1,14 @@
+mutation HolofuelRefundAllDeclined ($listOfDeclinedTransactions:[CancelDeclinedTransactionsInput]) {
```
Comment (robbiecarlton) [inline src/graphql/HolofuelRefundAllDeclinedMutation.gql]: see comment on `schema.graphql` about this

File: src/holofuel/components/RecentCounterparties/RecentCounterparties.js [hunk 15]
```diff
@@ -2,15 +2,20 @@ import React from 'react'
 import { isEmpty } from 'lodash/fp'
 import cx from 'classnames'
 import HashIcon from 'holofuel/components/HashIcon'
+import Loading from 'components/Loading'
 import { presentAgentId } from 'utils'
 import './RecentCounterparties.module.css'
 
-export default function RecentCounterparties ({ agents = [], selectedAgentId, selectAgent, className }) {
+export default function RecentCounterparties ({ agents = [], selectedAgentId, selectAgent, className, loading }) {
```
Comment (robbiecarlton) [inline src/holofuel/components/RecentCounterparties/RecentCounterparties.js:9]: weird that these changes are in this PR. They should already be in develop.

File: src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js [hunk 16]
```diff
@@ -24,13 +27,36 @@ export function PrimaryLayout ({
...
+
   const inboxCount = actionableTransactions.filter(actionableTx => actionableTx.status !== STATUS.canceled && !((actionableTx.status === STATUS.declined) && (actionableTx.type === TYPE.request))).length
 
   const isWide = useContext(ScreenWidthContext)
   const [isMenuOpen, setMenuOpen] = useState(false)
   const hamburgerClick = () => setMenuOpen(!isMenuOpen)
   const handleMenuClose = () => setMenuOpen(false)
 
+  const { push } = useHistory()
+  // eslint-disable-next-line
```
Comment (robbiecarlton) [inline src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js:41]: this is disabling a `missing dependencies' warning?

File: src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js [hunk 17]
```diff
@@ -24,13 +27,36 @@ export function PrimaryLayout ({
...
+  const goToInbox = useCallback(() => {
+    if (!hasCalledReroute) {
+      push(INBOX_PATH)
+      setHasCalledReroute(true)
+    }
+  })
```
Comment (robbiecarlton) [inline src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js:47]: I feel vary wary of this pattern of setting state to tell useEffect whether to work or not. I understand why you're doing it here but I think there's a better way. Just not sure what yet 😆 

I'm going to do some research today.

This is not blocking but I consider it code debt (I used the same in a recent PR, and marked it with a comment as temporary)
Can you add a TODO comment, something like "TODO: research a better way of doing this"

File: src/holofuel/pages/CreateOfferRequest/CreateOfferRequest.js [hunk 18]
```diff
@@ -245,14 +247,11 @@ export function RenderNickname ({ agentId, setCounterpartyNick, setCounterpartyF
   if (loading) {
     // TODO: Unsubscribe from Loader to avoid any potential mem leak.
     return <>
-      <Loader
+      <Loading
```
Comment (robbiecarlton) [inline src/holofuel/pages/CreateOfferRequest/CreateOfferRequest.js:250]: again, I'm confused why these changes are appearing in this PR when they're already on develop. /shrug

File: src/holofuel/pages/Inbox/Inbox.js [hunk 19]
```diff
@@ -80,29 +104,21 @@ function useCounterparty (agentId) {
   return { holofuelCounterparty, loading }
 }
 
-function useTransactionsWithCounterparties () {
-  const { data: { holofuelUser: whoami = {} } = {} } = useQuery(HolofuelUserQuery)
-  const { data: { holofuelInboxCounterparties = [] } = {} } = useQuery(HolofuelInboxCounterpartiesQuery, { fetchPolicy: 'cache-and-network' })
-  const { data: { holofuelActionableTransactions = [] } = {} } = useQuery(HolofuelActionableTransactionsQuery, { fetchPolicy: 'cache-and-network' })
-  const { data: { holofuelNonPendingTransactions = [] } = {} } = useQuery(HolofuelNonPendingTransactionsQuery, { fetchPolicy: 'cache-and-network' })
-
-  const updateCounterparties = (transactions, counterparties) => transactions.map(transaction => ({
```
Comment (robbiecarlton) [inline src/holofuel/pages/Inbox/Inbox.js:89]: 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 

Very happy that this is gone.

We should make a chore to do the same thing for the other Pages.

File: src/holofuel/pages/Inbox/Inbox.js [hunk 20]
```diff
@@ -80,29 +104,21 @@ function useCounterparty (agentId) {
...
-    ...transaction,
-    counterparty: counterparties.find(counterparty => counterparty.id === transaction.counterparty.id) || transaction.counterparty
-  }))
-
-  const allCounterparties = uniqBy('id', holofuelInboxCounterparties.concat([whoami]))
+function useUpdatedTransactionLists () {
```
Comment (robbiecarlton) [inline src/holofuel/pages/Inbox/Inbox.js:107]: I don't think we need this hook anymore. Better to have one hook for each query, that is responsible for whatever clean up we want to do. 

But that update can be part of the chore I suggested over here
<-----

File: src/holofuel/pages/Inbox/Inbox.js [hunk 21]
```diff
@@ -118,30 +134,51 @@ const presentTruncatedAmount = (string, number = 15) => {
...
+  useEffect(() => {
+    if (!isEmpty(filterActionableTransactionsByStatusAndType(STATUS.declined, TYPE.offer))) {
+      if (!hasDisplayedDeclinedTransactionsMessage) {
+        setIsDeclinedTransactionModalVisible(true)
+        setHasDisplayedDeclinedTransactionsMessage(true)
+      }
+    } else {
+      setHasDisplayedDeclinedTransactionsMessage(false)
+    }
+  }, [filterActionableTransactionsByStatusAndType, hasDisplayedDeclinedTransactionsMessage, setIsDeclinedTransactionModalVisible])
```
Comment (robbiecarlton) [inline src/holofuel/pages/Inbox/Inbox.js:160]: see my previous comment about this pattern.

File: src/holofuel/pages/Inbox/Inbox.test.js [hunk 22]
```diff
@@ -1,14 +1,14 @@
 import React from 'react'
+// import Modal from 'react-modal'
```
Comment (robbiecarlton) [inline src/holofuel/pages/Inbox/Inbox.test.js]: remove comment

File: src/holofuel/pages/Inbox/Inbox.test.js [hunk 23]
```diff
@@ -1,14 +1,14 @@
...
 import { fireEvent, act, within } from '@testing-library/react'
 import wait from 'waait'
 import moment from 'moment'
-import { reverse } from 'lodash/fp'
 import { ApolloProvider } from '@apollo/react-hooks'
 import { MockedProvider } from '@apollo/react-testing'
 import apolloClient from 'apolloClient'
-import Inbox, { TransactionRow, ConfirmationModal } from './Inbox'
+import Inbox, { TransactionRow, ConfirmationModal, DeclinedTransactionModal } from './Inbox' // , DeclinedTransactionModal
```
Comment (robbiecarlton) [inline src/holofuel/pages/Inbox/Inbox.test.js]: remove comment

**Review by robbiecarlton — CHANGES_REQUESTED** [review]

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 3 above)
Comment (JettTech) [reply src/data-interfaces/HoloFuelDnaInterface.js]: Works for me; but I do like how it coincides with the verbiage already used in the interface.

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 5 above)
Comment (JettTech) [reply src/data-interfaces/HoloFuelDnaInterface.js]: Noted. :+1:

File: src/holofuel/pages/CreateOfferRequest/CreateOfferRequest.js (hunk 18 above)
Comment (JettTech) [reply src/holofuel/pages/CreateOfferRequest/CreateOfferRequest.js:250]: Hm... That is strange. :thinking:

File: src/holofuel/components/RecentCounterparties/RecentCounterparties.js (hunk 15 above)
Comment (JettTech) [reply src/holofuel/components/RecentCounterparties/RecentCounterparties.js:9]: Agreed.

File: src/graphql/HolofuelAcceptOfferMutation.gql (hunk 13 above)
Comment (JettTech) [reply src/graphql/HolofuelAcceptOfferMutation.gql:5]: We need to use the nickname (or if not available, the id) when displaying the transaction we just edited and before the new `list_pending` data is returned.

File: src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js (hunk 17 above)
Comment (JettTech) [reply src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js:47]: Good feedback and thoughts.  I'd be happy to hear results from your recent research as well as collective consider a cleaner approach. :100:

File: src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js (hunk 16 above)
Comment (JettTech) [reply src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.js:41]: Exactly. `react-hooks/exhaustive-deps` :100: 

Not a standard, nor a preference, just apart of the unique setup and temporary solutions, as outlined below.

File: src/graphql-server/schema.graphql (hunk 11 above)
Comment (JettTech) [reply src/graphql-server/schema.graphql]: So, unfortunately, I learned the hard way we needed this. After some research, I found this solution read this one straight from the Apollo docs. - Exchanging this for a list of inline args does not though, as not all of the args are primitive types.

That said, please let me know if you have alternative knowledge or insights.

File: src/graphql-server/schema.graphql (hunk 12 above)
Comment (JettTech) [reply src/graphql-server/schema.graphql]: Naming works. I'll update that now. :+1:

As mentioned above, the list of args does not though, as not all of them are primitive types.  I actually researched this solution read this one straight from the Apollo docs.  Once again, if you have a different awareness, please don't hesitate to share and we'll discuss. :pray:

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 7 above)
Comment (JettTech) [reply src/data-interfaces/HoloFuelDnaInterface.js]: Ah, I believe this was an old `console.error` that  was updated in a hurry.

Nonetheless, this is a valid point about the `canceledProof`.
Good catch!

File: src/graphql/HolofuelRefundAllDeclinedMutation.gql (hunk 14 above)
Comment (JettTech) [reply src/graphql/HolofuelRefundAllDeclinedMutation.gql]: I'll update the name here. :+1:

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 7 above)
Comment (robbiecarlton) [reply src/data-interfaces/HoloFuelDnaInterface.js]: Please update

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 8 above)
Comment (robbiecarlton) [reply src/data-interfaces/HoloFuelDnaInterface.js]: same

File: src/data-interfaces/HoloFuelDnaInterface.js (hunk 9 above)
Comment (robbiecarlton) [reply src/data-interfaces/HoloFuelDnaInterface.js]: please rename

File: src/graphql-server/schema.graphql (hunk 11 above)
Comment (robbiecarlton) [reply src/graphql-server/schema.graphql]: We only need to pass in the ids. See comment on `holofuelRefundAllDeclined` below.

File: src/graphql-server/schema.graphql (hunk 12 above)
Comment (robbiecarlton) [reply src/graphql-server/schema.graphql]: Let's jump on zoom to resolve this. Seems we're missing each other.

File: src/graphql/HolofuelRefundAllDeclinedMutation.gql (hunk 14 above)
Comment (robbiecarlton) [reply src/graphql/HolofuelRefundAllDeclinedMutation.gql]: please update the name

File: src/holofuel/pages/Inbox/Inbox.test.js (hunk 22 above)
Comment (robbiecarlton) [reply src/holofuel/pages/Inbox/Inbox.test.js]: please remove comment

File: src/graphql-server/schema.graphql [hunk 24]
```diff
@@ -112,7 +112,7 @@ input HolofuelUserInput {
   nickname: String
 }
 
-input CancelDeclinedTransactionsInput {
+input TransactionsInput {
```
Comment (robbiecarlton) [inline src/graphql-server/schema.graphql]: please change to `TransactionInput` (not plural)

**Review by robbiecarlton — CHANGES_REQUESTED** [review]

File: src/graphql-server/schema.graphql (hunk 24 above)
Comment (JettTech) [reply src/graphql-server/schema.graphql]: Done.

**Review by zo-el — COMMENTED** [review]
Looks good :+1:
